File: 11/11279.cpp
/*
priority queue
*/

#define _CRT_SECURE_NO_WARNINGS

#include <iostream>
#include <queue>
using namespace std;

int main() {
	int N; scanf("%d", &N);	// 연산의 개수
	priority_queue<int> heap;	// 최대 힙

	for (int i = 0; i < N; i++) {
		int x; scanf("%d", &x);

		if (x == 0) {
			if (heap.empty()) printf("0\n");
			else {
				printf("%d\n", heap.top());
				heap.pop();
			}
		}
		else heap.push(x);
	}
}

File: 11/11279.py
import queue
import sys

if __name__ == "__main__":
    N = int(sys.stdin.readline())
    heap = queue.PriorityQueue()

    for i in range(N):
        x = int(sys.stdin.readline())

        if x == 0:
            if heap.empty(): sys.stdout.write("0\n")
            else: sys.stdout.write(f"{-heap.get()}\n")
        else: heap.put(-x)
